### PR TITLE
experiments/colgranule: add reusable granule core

### PR DESCRIPTION
**Pull request by snissn**
## Objective

Implement Phase 1 / PR1 from #1534: a focused, high-performance column granule writer/reader core in `experiments/colgranule`.

## Context

This keeps column store work moving independently of WAL/durable TreeDB integration. The PR establishes the inner-loop API for writing int64 granules, decoding them through reusable scratch, validating metadata, and measuring encode/decode/scan throughput and allocation behavior.

Refs #1534.

## Non-goals

- No durable on-disk part format.
- No TreeDB root/WAL integration.
- No generational garbage collection.
- No multi-column execution engine.
- No generic typed-vector framework beyond the current int64 lane.

## Design

- Expands `EncodedGranule` with row/null/default counts, stored bytes, and inline payload refs.
- Adds `GranuleBuilder` for reusable-scratch int64 granule construction.
- Adds `GranuleReader` for reusable-scratch decode and range-count scan paths.
- Keeps existing `EncodeInt64`, `DecodeInt64`, and `RangeScanCount` entry points for current callers.
- Validates row counts, null/default counts, stored payload lengths, payload ref lengths, raw int64 length divisibility, and decompressed lengths before decoding.
- Keeps corrupt/truncated payloads fail-closed with returned errors.

## Scope

- Includes (files/symbols): `experiments/colgranule/granule.go`, `GranuleBuilder`, `GranuleReader`, `EncodedGranule`, int64 encode/decode/range count, JSONBench colgranule summary path, colgranule tests and benchmarks.
- Excludes (files/symbols): TreeDB production storage, WAL, disk part layout, GC, non-int64 column types.

## Correctness

- Tests run (exact commands):
  - `go test ./experiments/colgranule`
  - `go test ./experiments/colgranule -run '^$' -bench 'Benchmark(Encode|Decode|RangeScan)Int64Granule' -benchmem -benchtime=100x`
  - `go test ./experiments/colgranule -run TestCompressionRatios -v`
  - `go test ./...` (one unrelated `TreeDB/caching` test failed once; targeted rerun passed)
  - `go test ./TreeDB/caching -run '^TestCheckpoint_KicksVlogGenerationRewriteDespiteRecentForegroundActivity$' -count=1`
- Corruption/edge cases covered: empty granules rejected, min/max exactness for negative/positive/repeated/monotonic values, truncated raw payloads, truncated/trailing delta payloads, truncated Snappy payloads, invalid stored-byte metadata, reused decode buffers returning the correct shorter result.
- Concurrency/race coverage (if applicable): not applicable; this PR adds per-builder/per-reader scratch structs with no shared package-global mutable state.

## Performance

- Benchmarks run (exact commands): `go test ./experiments/colgranule -run '^$' -bench 'Benchmark(Encode|Decode|RangeScan)Int64Granule' -benchmem -benchtime=100x`
- Environment: `darwin/arm64`, Apple M3, 8192-row granules.
- Allocation gate: the full warmed sweep reported `0 B/op` and `0 allocs/op` for encode, decode, range-hit, and min/max-skip benchmark variants.
- Before/after: median-of-5 before/after is not available because this is a new experiment API surface; this PR adds the benchmark coverage used below.

Representative results from the full sweep:

| Operation | Fixture | Encoding / compression | ns/row | rows/s | MiB/s | stored B/row | B/op | allocs/op |
| --- | --- | --- | ---: | ---: | ---: | ---: | ---: | ---: |
| Encode | `low_cardinality` | `raw_int64` / `none` | 1.242 | 805.3M | 6144 | 8.000 | 0 | 0 |
| Encode | `low_cardinality` | `delta_varint` / `lz4` | 3.218 | 310.8M | 2371 | 0.008667 | 0 | 0 |
| Encode | `random` | `delta_varint` / `snappy` | 7.092 | 141.0M | 1076 | 9.240 | 0 | 0 |
| Decode | `monotonic` | `raw_int64` / `none` | 0.5887 | 1.699B | 12959 | 8.000 | 0 | 0 |
| Decode | `low_cardinality` | `raw_int64` / `lz4` | 1.251 | 799.6M | 6101 | 0.1260 | 0 | 0 |
| Decode | `random` | `delta_varint` / `snappy` | 6.687 | 149.6M | 1141 | 9.240 | 0 | 0 |
| Range hit | `low_cardinality` | `delta_varint` / `snappy` | 1.939 | 515.7M | 3934 | 0.04785 | 0 | 0 |
| Range hit | `random` | `delta_varint` / `snappy` | 7.933 | 126.1M | 961.7 | 9.240 | 0 | 0 |

Additional notes:

- Min/max skip paths are metadata-only and measured around 4 to 5 ns/op with `0 B/op` and `0 allocs/op` in this run.
- Worst-case incompressible fixtures are included via `random`; LZ4 correctly falls back to `actual_none` for incompressible random payloads where compression would not shrink data.
- Regressions (if any) and why acceptable: none known in the focused colgranule package.

## Rollout / Toggle Plan

- Default setting: experiment-only API; no production behavior is enabled.
- How to disable quickly: revert this PR or avoid the `experiments/colgranule` package.

## Follow-ups

- Issues/PRs to do next: continue #1534 Phase 1 with PR2 for typed column batches and multi-column granule sets after PR1 lands.

---

### Scope
- [x] Single, clear objective for this PR (not a bundle)
- [x] Explicit include list (files/symbols touched): `experiments/colgranule/granule.go`, `granule_test.go`, `granule_bench_test.go`, `jsonbench_report.go`
- [x] Explicit exclude list (files/symbols NOT touched): TreeDB production storage, WAL, durable part layout, GC
- [x] No unwanted feature reintroduction (e.g. async slab writer, hard zones, ValueIndex) unless explicitly stated in the PR objective

### Safety / Correctness
- [x] Clear written-vs-durable semantics for any `*Sync` path touched: not applicable; no `*Sync` path touched
- [x] No new mmap usage on mutable/truncating files
- [x] All new length fields are capped before allocation (OOM-safe)
- [x] CRC/checksum behavior is fail-closed (clean error, no panic): not applicable to CRC, but validation paths are fail-closed
- [x] Any concurrency change has a defined state machine and invariants: not applicable; no concurrency change

### Tests
- [x] Added deterministic regression tests for the targeted failure mode
- [x] Tests avoid sleeps where possible (use latches/hooks)
- [x] Added corruption/edge-case tests (truncation, bad headers, invalid lengths) when format/parsing changes
- [x] Ran locally:
  - [ ] `go test ./... -count=1`
  - [x] Any targeted packages/benchmarks: commands listed above

### Benchmarks / Measurements
- [x] Added or updated a benchmark relevant to the change
- [x] Reported results in PR description (before/after, command, machine)
- [x] Included “incompressible” (worst-case) results if compression is involved

### Docs
- [x] No repo doc update needed for this experiment-only API
- [x] No new config knobs
- [x] Called out any format change in PR description (pre-alpha OK)

### Rollout / Toggle Plan (if behavior-affecting)
- [x] Safe defaults (feature off or conservative threshold)
- [x] Clear knobs to disable/revert behavior quickly
- [x] Observability: not applicable; no production behavior enabled

## Summary by CodeRabbit

* **Refactor**
  * Improved encode/decode pipeline for faster, more memory-efficient reads/writes and stronger integrity checks to detect corrupt or inconsistent data.

* **Benchmarks**
  * Updated benchmarks and reporting to measure stored-byte metrics and clarify subtest names, reflecting realistic compression and throughput.

* **Tests**
  * Expanded test coverage for empty/oversized inputs, metadata invariants, compression round-trips, buffer-reuse safety, and corruption rejection.

[Review Change Stack](https://app.coderabbit.ai/change-stack/snissn/gomap/pull/1537?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: experiments/colgranule/granule.go [hunk 1]
```diff
@@ -193,6 +313,40 @@ func decodeDeltaVarint(dst []int64, raw []byte, rows int) ([]int64, error) {
...
+	case CompressionNone:
+		return raw, CompressionNone, dst[:0], nil
+	case CompressionSnappy:
+		need := snappy.MaxEncodedLen(len(raw))
+		if cap(dst) < need {
+			dst = make([]byte, need)
+		} else {
+			dst = dst[:need]
+		}
+		out := snappy.Encode(dst, raw)
```
[the remark this answers is not in the document]
Comment (snissn) [reply experiments/colgranule/granule.go:352]: Fixed in d636f3342: `compressPayloadInto` now passes a zero-length scratch slice to `snappy.Encode`, with a regression test that decodes a snappy-built granule.

File: experiments/colgranule/granule.go [hunk 2]
```diff
@@ -169,6 +286,9 @@ func encodeInt64Payload(dst []byte, values []int64, encoding Encoding) ([]byte,
 
 func decodeDeltaVarint(dst []int64, raw []byte, rows int) ([]int64, error) {
 	out := dst[:0]
+	if cap(out) < rows {
+		out = make([]int64, 0, rows)
+	}
```
[the remark this answers is not in the document]
Comment (snissn) [reply experiments/colgranule/granule.go:306]: Fixed in fdc8013fb: delta decode now validates row bounds and minimum payload bytes before preallocating, with a huge-rows/tiny-payload corrupt regression test.

File: experiments/colgranule/granule.go [hunk 3]
```diff
@@ -193,6 +322,50 @@ func decodeDeltaVarint(dst []int64, raw []byte, rows int) ([]int64, error) {
 	return out, nil
 }
 
+func validateGranuleDecodeRows(rows int) error {
+	if rows < 0 {
+		return fmt.Errorf("colgranule: negative rows=%d", rows)
+	}
+	if rows > maxGranuleDecodeRows {
+		return fmt.Errorf("colgranule: rows=%d exceed cap %d", rows, maxGranuleDecodeRows)
+	}
```
[the remark this answers is not in the document]
Comment (snissn) [reply experiments/colgranule/granule.go:337]: Fixed in ccc5fe1f3: BuildInt64 and EncodeInt64 now reject row counts above the same cap enforced by decode, so new oversized granules fail before write instead of becoming non-round-trippable.

File: experiments/colgranule/granule.go [hunk 4]
```diff
@@ -193,6 +322,50 @@ func decodeDeltaVarint(dst []int64, raw []byte, rows int) ([]int64, error) {
...
+		return raw, CompressionNone, dst[:0], nil
+	case CompressionSnappy:
+		need := snappy.MaxEncodedLen(len(raw))
+		if cap(dst) < need {
+			dst = make([]byte, need)
+		} else {
+			dst = dst[:0]
+		}
+		out := snappy.Encode(dst, raw)
+		return out, CompressionSnappy, out, nil
```
[the remark this answers is not in the document]
Comment (snissn) [reply experiments/colgranule/granule.go:353]: Fixed in ccc5fe1f3: the snappy builder scratch path now allocates zero-length capacity before snappy.Encode, and the existing round-trip regression covers the builder path.

File: experiments/colgranule/granule.go [hunk 5]
```diff
@@ -248,6 +444,34 @@ func decompressPayload(g EncodedGranule) ([]byte, error) {
...
+	}
+	if g.NullCount > g.Rows || g.DefaultCount > g.Rows-g.NullCount {
+		return errors.New("colgranule: null/default count exceeds rows")
+	}
+	if g.RawBytes < 0 {
+		return fmt.Errorf("colgranule: negative raw payload length %d", g.RawBytes)
+	}
+	if g.StoredBytes < 0 {
+		return fmt.Errorf("colgranule: negative stored payload length %d", g.StoredBytes)
+	}
```
[the remark this answers is not in the document]
Comment (snissn) [reply experiments/colgranule/granule.go:474]: Fixed in ccc5fe1f3: validateGranule now applies the row cap before decode allocation and bounds raw/stored/payload lengths by the encoding and compression shape.

File: experiments/colgranule/granule.go [hunk 6]
```diff
@@ -248,6 +444,34 @@ func decompressPayload(g EncodedGranule) ([]byte, error) {
...
+	if g.StoredBytes != 0 && g.StoredBytes != len(g.Payload) {
+		return fmt.Errorf("colgranule: stored payload length=%d want=%d", len(g.Payload), g.StoredBytes)
+	}
+	if g.PayloadRef.Length != 0 && g.PayloadRef.Length != len(g.Payload) {
+		return fmt.Errorf("colgranule: payload ref length=%d want=%d", g.PayloadRef.Length, len(g.Payload))
+	}
```
[the remark this answers is not in the document]
Comment (snissn) [reply experiments/colgranule/granule.go]: Fixed in ccc5fe1f3: inline payload refs now reject unknown kinds, non-zero offsets, and mismatched StoredBytes/PayloadRef.Length even when the bad value is zero.
